src/main/java/org/spongepowered/api/text/template/TextTemplate.java

```diff
@@ -0,0 +1,214 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.text.template;
+
+import org.spongepowered.api.text.Text;
+import org.spongepowered.api.text.TextElement;
+import org.spongepowered.api.text.channel.MessageReceiver;
+
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Represents a re-usable template that produces a formatted
+ * {@link Text.Builder}. Elements will be appended to the result builder in the
+ * order that they are specified in {@link #of(Object...)}.
+ *
+ * TODO: Serialization
+ */
+public class TextTemplate {
+    final Object[] elements;
+
+    TextTemplate(Object... elements) {
+        this.elements = elements;
+    }
+
+    /**
+     * Evaluates this TextTemplate and returns the result in a
+     * {@link Text.Builder}.
+     *
+     * @param params Parameters to evaluate
+     * @return Text builder containing result
+     * @throws RequiredTextTemplateArgumentException if required parameters are missing
+     */
+    public Text.Builder eval(Map<String, Object> params) {
+        Text.Builder result = Text.builder();
+        for (Object element : this.elements) {
+            appendElement(result, element, params);
+        }
+        return result;
+    }
+
+    private void appendElement(Text.Builder builder, Object element, Map<String, Object> params) {
+        if (element instanceof TextElement) {
+            ((TextElement) element).appendTo(builder);
+        } else if (element instanceof Arg) {
+            Arg<?> arg = (Arg<?>) element;
+            Object param = params.get(arg.name);
+            if (param == null) {
+                if (!arg.optional) {
+                    throw new RequiredTextTemplateArgumentException(arg.name);
+                }
+            } else {
+                arg.eval(param).appendTo(builder);
+            }
+        } else {
+            builder.append(Text.of(element.toString()));
+        }
+    }
+
+    /**
+     * Sends the result of this TextTemplate with the specified parameters to
+     * the specified {@link MessageReceiver}.
+     *
+     * @param receiver Receiver to send message to
+     * @param params Parameters to apply to template
+     */
+    public void send(MessageReceiver receiver, Map<String, Object> params) {
+        receiver.sendMessage(eval(params).build());
+    }
+
+    /**
+     * Constructs a new TextTemplate for the given elements. The order of the
+     * elements is the order in which they will be appended to the result
+     * builder via {@link #eval(Map)}.
+     *
+     * The provided elements may be of any type.
```

Comment:
There should be `<p>` `</p>` around all extra paragraphs (so they actual get displayed as paragraphs)